### ISSUE TITLE
High CPU on SQL requests after Azure SQL Server switch (eg. maintenance, upgrade)

### ISSUE THREAD

**Issue by gluwer** [tediousjs/tedious#1136]
After Azure SQL database switch (because of internal upgrade of SQL engine/VM or configuration change like update from S2 to S3), we observe a CPU used spike after 3 to 10 minutes after the switch. If the connections are dropped and new ones are created, everything is going back to normal. If there will be only old connections, the high cpu consumption per request will stay for hours (it is not just a bum parter the switch).

The amount of CPU spike is proportional to SQL request traffic, for example if database CPU was 5% it goes up to 50%, not 100%. All requests are still working properly and return the data as they should.

I've attached the script (https://gist.github.com/gluwer/8556ac515de53e497f611d0984627b86) that allows to test the situtation, but it requires creating a simple Azure SQL DB. It was tested on S0 and S1, but we observe the same CPU spikes on P2 databases too. Script is a bit dummy, but allows to reproduce the issue easily. Script contains only CRUD operations, and no stored procedures.
Below is a log from one of the runs of it:
```
100
5 5
Database cpu 60 (3.923)
...
Database cpu 60 (18.817)
1100
1200
6 5
Database cpu 60 (18.479)
1300
1400
6 5
Database cpu 60 (18.43)
Pool error (bint_test) ESOCKET
Error: read ECONNRESET
...
Ended connection
40613 : Database 'test' on server 'test' is not currently available.  Please retry the connection later.  If the problem persists, contact customer support, and provide them the session tracing ID of '333333'.
...
1600
...
Database cpu 60 (36.05125)
2400
2500
6 5
Database cpu 60 (35.611)
...
Database cpu 60 (35.411)
4800
6 5
Database cpu 60 (48.291)
6 5
Database cpu 60 (61.44)
Cpu above threshold, use new connection pool
Database bint_test new pool started
Ended connection
Ended connection
Ended connection
Ended connection
Ended connection
Ended connection
Database bint_test oldPool close
...
5600
5 5
Database cpu 60 (52.196)
5701 : Changed database context to 'bint_test'.
5703 : Changed language setting to us_english.
5703 : Changed language setting to us_english.
5700
5800
6 5
Database cpu 60 (39.065)
5900
6000
6 5
Database cpu 60 (35.522)
6100
6 5
Database cpu 60 (34.978)
6300
6400
6 5
Database cpu 60 (33.996)
```

The problem occures almost always. We have used newest tedious 9.0.1 and nodejs 12.16.3.

Below is how it all looks from database perspective:

```
end_time	avg_cpu_percent	avg_data_io_percent	avg_log_write_percent	avg_memory_usage_percent	xtp_storage_percent	max_worker_percent	max_session_percent
2020-08-06 09:15:21.760	99.78	0.00	16.01	1.50	0.00	6.66	1.33
2020-08-06 09:15:06.750	99.51	0.00	18.10	1.49	0.00	3.33	1.33
2020-08-06 09:14:51.733	99.65	0.00	15.90	1.49	0.00	6.66	1.33
2020-08-06 09:14:36.727	80.05	0.00	28.06	1.49	0.00	8.33	1.33
2020-08-06 09:14:21.680	33.85	0.00	56.81	1.49	0.00	8.33	1.33
2020-08-06 09:14:06.683	33.88	0.00	54.06	1.48	0.00	8.33	1.33
```

When there is only one connection and cpu bumps only to 50% for example, all other stats stay the same, only cpu jumps.

What we have checked looking for a possible solution:

* tedious debug and requests content (before and when there is high cpu all packets look exactly the same including state switches reported etc.)
* tedious protocol and some other options (checked 7.3 and 7.2, but problem still exists there)
* database comptatibility level (from 100 to 140, no effect)
* connections don't have to exist when there is a switch, eg. you can make a switch and start new conections seconds later and the high cpu still occurs (but if you connect after 20 mins, everything will be fine)
* there can be only one connection with some load that will make cpu go from 5% to 50% on S0 database
* doing a second switch when first one high cpu occures does not make it happen again even after 20 minutes
* database still reports proper number of connections, they are not doubled etc.
* reported in log analytics query cpu is the same as before, but query duration jumps (normally both are very similar)
* when high cpu, there is much more SOS_SCHEDULER wait type than normally

We have also checked what happens when if we use .net entity framework with similar load and do the switch - there is no high cpu within next 20 minutes. At the beginning we have reported the issue on mssql (https://github.com/tediousjs/node-mssql/issues/1067), as we do not use tedious directly.

**Conslusions**

We at the moment do not know if the problem is related to node, tedious, MS SQL server or the Azure SQL Gateway.
It looks like gateway or SQL server is doing something minutes after the switch (maybe on low level) which makes this strange high cpu spike.

**Comment by arthurschreiber** [comment]
> * connections don't have to exist when there is a switch, eg. you can make a switch and start new conections seconds later and the high cpu still occurs (but if you connect after 20 mins, everything will be fine)

Does this mean that if you perform the switch, and then connect, you see the high load? That points to this not being an issue caused by `tedious`, but a server side problem. 🤔 

@IanChokS @David-Engel @MichaelSun90 FYI.

**Comment by gluwer** [comment]
Yes, I understand the logic and this may be the case, but why it does not happen when we do similar load and switch using .net entity framework client? Maybe sql server does something that tedious or nodejs is not handling the way it expects.

I'll clarify the quoted situation. There is a tedious load on the DB, we make a switch, connections are dropped as usual (socket reset), new ones are created (after the database unavailable texts), we stop the node proccess when all is up and then starts it again, and the high cpu will still happen several minutes later. If the DB is idle when we make the switch and then start nodejs/tedious, there **won't be** any high cpu.

This is why it may be the bug is sql server somewhere, but likely nodejs/tedious is taking a role in it, but we do not know how.

**Comment by David-Engel** [comment]
@gluwer Have you opened a case with Azure support? I think this might be easier for them to troubleshoot from the server side and seeing what exactly the server-side process is doing that is causing the high CPU load. That might help identify something tedious is doing incorrectly (if anything) or a bug on the server-side.

I don't see anything above that indicates you are getting any client-side errors or incorrect behavior, so I would guess that this is a server-side issue.

**Comment by gluwer** [comment]
@David-Engel We still haven't, because currently we are in the middle of subscription migration to the new partner at the moment and earlier we were more on the lib cause side because of not seeing it on .net client. When we finalize the migration, we are going to open a ticket.

Yes, you are right. When high cpu happens, queries and everything else works exactly the same except very high reported cpu which goes back to normal levels after dropping old connections and starting new ones (without process reset).

**Comment by gluwer** [comment]
We decided to rise a ticket using old subscription, as issue happened on production again this Wednesday.

**Comment by gluwer** [comment]
Hmm, the high cpu after the switch seems to be gone on our test database today. The only noticed change is that database engine reported by `select @@version` was updated from Microsoft SQL Azure (RTM) - 12.0.2000.8 Jun 18 2020 16:54:11 to Microsoft SQL Azure (RTM) - 12.0.2000.8 Jul 31 2020 08:26:29.

The support we contacted about the issue is not convinced this should be the reason the problem is gone, but as the problem started to be visible in the beginning of July (likely introduction of previous engine build) and with new version seems to be gone, DB engine bug theory is plausible now.

@David-Engel Is there any publically available changelog of what was fixed in Azure SQL like for normal SQL Server 2019 (https://sqlserverupdates.com/sql-server-2019-updates/)?
